src/backend/aspen/api/utils/phylo.py

```diff
@@ -60,6 +90,86 @@ def _sample_filter(sample: Sample, can_see_pi_group_ids: Set[int], system_admin:
     return sample.submitting_group_id in can_see_pi_group_ids
 
 
+def _collect_countries(node: dict) -> Set[str]:
+    countries = set()
+    node_country = node.get("node_attrs", {}).get("country", {}).get("value", None)
+    if node_country:
+        countries.add(node_country)
+    for child in node.get("children", []):
+        countries |= _collect_countries(child)
```

Comment:
is this a function calling itself 🤯

Reply:
Yep, it recursively traverses the nodes in the tree.